cmd/snaplock/runinhibit/inhibit.go

```diff
@@ -283,3 +278,113 @@ func RemoveLockFile(snapName string) error {
 
 	return nil
 }
+
+// needed for better mocking
+type ticker interface {
+	Wait()
+	Stop()
+}
+
+type timeTicker struct {
+	*time.Ticker
+}
+
+func (t *timeTicker) Wait() {
+	if t.Ticker != nil {
+		<-t.Ticker.C
+	}
+}
+
+var newTicker = func(interval time.Duration) ticker {
+	return &timeTicker{time.NewTicker(interval)}
+}
+
+// WaitWhileInhibited blocks until snap is not inhibited anymore returning
+// a locked hint file lock.
+//
+// "notInhibted" and "inhibited" callbacks are called with the hint file lock held.
+// If inhibited callback returns true, WaitWhileInhibited returns instantly without
+// waiting for the snap not being inhibited.
+//
+// NOTE: It is the caller's responsibility to release the returned file lock.
+var WaitWhileInhibited = func(snapName string, notInhibited func() error, inhibited func(hint Hint, inhibitInfo *InhibitInfo) (cont bool, err error), interval time.Duration) (flock *osutil.FileLock, retErr error) {
+	ticker := newTicker(interval)
+	defer ticker.Stop()
+
+	if err := os.MkdirAll(InhibitDir, 0755); err != nil {
+		return nil, err
+	}
+
+	for {
+		flock, err := openHintFileLock(snapName)
+		if err != nil {
+			return nil, err
+		}
+		// release lock if we return early with an error
+		defer func() {
+			// keep lock if no errors occur
+			if retErr != nil {
+				flock.Close()
+			}
+		}()
+
+		// hold read lock
+		if err := flock.ReadLock(); err != nil {
+			return nil, err
+		}
+
+		// read inhibition hint
+		hint, err := hintFromFile(flock.File())
+		if err != nil {
+			return nil, err
+		}
+
+		if hint == HintNotInhibited {
+			if notInhibited != nil {
+				if err := notInhibited(); err != nil {
+					return nil, err
+				}
+			}
+			return flock, nil
+		} else {
+			if inhibited != nil {
+				inhibitInfo, err := readInhibitInfo(snapName, hint)
+				if err != nil {
+					return nil, err
+				}
+				cont, err := inhibited(hint, &inhibitInfo)
+				if err != nil {
+					return nil, err
+				}
+				if cont {
+					return flock, nil
+				}
+			}
+		}
+
+		// close the lock file explicitly so we can try to lock again after waiting for the interval.
+		flock.Close()
+		ticker.Wait()
```

Comment:
Great catch, Thank you! Do you think switching to sleep instead of a ticker would be more reasonable?

Reply:
I think sleep would be a bit safer, yeah, just to guarantee there's actually time for it to possibly become uninhibited. It might make sense to scale this based on architecture (see `testutil.HostScaledTimeout`), but that's only if meaningful work has to be done with it unlocked. If the work to uninhibit the snap just involves waiting to acquire the lock, then take all the time it needs, then release, there should be no reason to scale the time slept here.

Hope this makes sense..... Basically I think sleep should probably be fine, and no need to scale sleep time based on architecture.